lib/string_decoder.js

```diff
@@ -279,6 +286,18 @@ function simpleWrite(buf) {
   return buf.toString(this.encoding);
 }
 
-function simpleEnd(buf) {
-  return (buf && buf.length ? this.write(buf) : '');
+function simpleEnd(ctx, buf) {
+  return (buf && buf.length ? ctx.write(buf) : '');
+}
+
+const endMappings = new Map([
+  ['utf16le', utf16End],
+  ['utf8', utf8End],
+  ['base64', base64End]
+]);
+
+function end(buf) {
+  const result = (endMappings.get(this.encoding) || simpleEnd)(this, buf);
```

Comment:
This looks "elegant" but using a switch case for the encoding and calling the function directly is probably faster.